File: Models/Match.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace WebQuanLyGiaiDau_NhomTD.Models
{
    public class Match
    {
        public int Id { get; set; }

        [Required]
        public string TeamA { get; set; }

        // ID của đội A (tùy chọn)
        public int? TeamAId { get; set; }

        [Required]
        public string TeamB { get; set; }

        // ID của đội B (tùy chọn)
        public int? TeamBId { get; set; }

        // Add score properties
        public int? ScoreTeamA { get; set; }
        public int? ScoreTeamB { get; set; }

        // Vòng đấu (ví dụ: 1, 2, 3, ...)
        public int? Round { get; set; }

        // Tên bảng đấu (ví dụ: "Bảng A", "Bảng B", ...)
        public string? GroupName { get; set; }

        [DataType(DataType.Date)]
        [Display(Name = "Ngày Thi Đấu")]
        public DateTime MatchDate { get; set; }

        // Thời gian thi đấu
        [Display(Name = "Thời Gian Thi Đấu")]
        [DataType(DataType.Time)]
        public TimeSpan? MatchTime { get; set; } = new TimeSpan(15, 0, 0); // Default to 15:00 (3 PM)

        // Địa điểm thi đấu
        [Display(Name = "Địa Điểm Thi Đấu")]
        public string? Location { get; set; }

        // YouTube video URLs for highlights and live stream
        [Display(Name = "Video Highlights")]
        public string? HighlightsVideoUrl { get; set; }

        [Display(Name = "Live Stream URL")]
        public string? LiveStreamUrl { get; set; }

        [Display(Name = "Mô tả Video")]
        public string? VideoDescription { get; set; }

        // Match status: "Upcoming", "InProgress", "Completed"
        // This property doesn't exist in the database
        [NotMapped]
        public string? Status { get; set; }

        // Calculate status based on match date
        [NotMapped] // This property is not stored in the database
        public string CalculatedStatus
        {
            get
            {
                if (MatchDate < DateTime.Now)
                    return "Completed";
                else if (MatchDate.Date == DateTime.Now.Date)
                    return "InProgress";
                else
                    return "Upcoming";
            }
        }

        // Foreign Key
        public int TournamentId { get; set; }

        // Navigation
        public Tournament Tournament { get; set; }
        public ICollection<Statistic>? Statistics { get; set; }
        public ICollection<MatchSet>? MatchSets { get; set; }
    }
}

File: tournament_app/lib/models/match.dart
import 'package:json_annotation/json_annotation.dart';

part 'match.g.dart';

@JsonSerializable()
class Match {
  @JsonKey(fromJson: _safeIntFromJson)
  final int id;
  @JsonKey(fromJson: _safeStringFromJson)
  final String teamA;
  @JsonKey(fromJson: _safeStringFromJson)
  final String teamB;
  final DateTime matchDate;
  final String? matchTime;
  final String? location;
  @JsonKey(fromJson: _safeIntFromJsonNullable)
  final int? scoreTeamA;
  @JsonKey(fromJson: _safeIntFromJsonNullable)
  final int? scoreTeamB;
  final String? groupName; // Bảng đấu (e.g., "Bảng A", "Bảng B")
  @JsonKey(fromJson: _safeIntFromJsonNullable)
  final int? round; // Vòng đấu (1, 2, 3, ...)

  Match({
    required this.id,
    required this.teamA,
    required this.teamB,
    required this.matchDate,
    this.matchTime,
    this.location,
    this.scoreTeamA,
    this.scoreTeamB,
    this.groupName,
    this.round,
  });

  static int _safeIntFromJson(dynamic value) {
    if (value == null) return 0;
    if (value is int) return value;
    if (value is double) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }

  static int? _safeIntFromJsonNullable(dynamic value) {
    if (value == null) return null;
    if (value is int) return value;
    if (value is double) return value.toInt();
    if (value is String) return int.tryParse(value);
    return null;
  }

  static String _safeStringFromJson(dynamic value) {
    if (value is String) return value;
    if (value is int) return value.toString();
    return value?.toString() ?? '';
  }

  factory Match.fromJson(Map<String, dynamic> json) => _$MatchFromJson(json);
  Map<String, dynamic> toJson() => _$MatchToJson(this);

  String get status {
    if (scoreTeamA != null && scoreTeamB != null) {
      return 'completed';
    } else if (matchDate.isBefore(DateTime.now())) {
      return 'ongoing';
    } else {
      return 'upcoming';
    }
  }

  String? get winner {
    if (scoreTeamA == null || scoreTeamB == null) return null;
    final scoreA = scoreTeamA!;
    final scoreB = scoreTeamB!;
    if (scoreA > scoreB) return teamA;
    if (scoreB > scoreA) return teamB;
    return 'draw';
  }

  @override
  String toString() => 'Match(id: $id, $teamA vs $teamB)';
}
